File: algorithms/security/php/code-response/src/RainbowTable/RainbowTable.php
<?php

namespace RainbowTable;

function array_sampling($chars, $size, $combinations = [])
{
  if (empty($combinations)) {
    $combinations = $chars;
  }

  if ($size == 1) {
    return $combinations;
  }

  $new_combinations = [];

  foreach ($combinations as $combination) {
    foreach ($chars as $char) {
      $new_combinations[] = $combination . $char;
    }
  }

  return array_sampling($chars, $size - 1, $new_combinations);
}

class RainbowTable
{
  public static function table($length)
  {
    $table = [];
    $letters = \str_split('abcdefghijklmnopqrstuvwxyz');
    foreach (array_sampling($letters, $length) as $combination) {
      $table[\md5($combination)] = $combination;
    }
    return $table;
  }

  public static function cracker($hash, $length)
  {
    $table = self::table($length);
    return $table[$hash];
  }
}

File: algorithms/security/ruby/code-response/src/RainbowTable/rainbow_table.rb
# use MD5.rb
# use array_combination.rb
require 'digest'

class RainbowTable
  attr_reader :table_data
  attr_reader :letters
  
  def initialize(length)
    @letters = ('a'..'z').to_a
    @table_data = self.table(length)
  end 

  def table(length)
  
    table_hash = {}
    letters
      .repeated_combination(length)
      .each do |combination| 
        table_hash[Digest::MD5.hexdigest combination.join('')] = combination.join('');
      end
    @table_data = table_hash  
      
  end

  def cracker(hash)
    @table_data[hash]
  end

end
